src/Sparrow/Json/BlittableJsonDocumentBuilder.cs

```diff
@@ -9,15 +10,21 @@
 
 namespace Sparrow.Json
 {
-    public sealed class BlittableJsonDocumentBuilder : AbstractBlittableJsonDocumentBuilder
+    public sealed unsafe class BlittableJsonDocumentBuilder : AbstractBlittableJsonDocumentBuilder
     {
         private static readonly StringSegment UnderscoreSegment = new StringSegment("_");
 
         private readonly JsonOperationContext _context;
         private UsageMode _mode;
         private readonly IJsonParser _reader;
-
         public IBlittableDocumentModifier _modifier;
+
+        // We introduced a delegate that encapsulates the reading logic.
+        // This avoids branching on every 'Read()' call to figure out which path to take
+        // (WriteNone vs. WriteFull, ObjectJsonParser vs. UnmanagedJsonParser).
+        // Instead, we decide once in the constructor or Renew(...) and store that strategy here.
+        private Func<bool> _readInternalFunc;
```

Comment:
A func pointer wouldn't be better?

Reply:
Nope in this case, that would push me to include too many references (remember they are static) and read internal function requires lots of state... there is no difference in performance at this level at the caller level.